[Filter,V2] Added a framework to support TF2-Lite (2.3.0) @open sesame 09/04 12:07

leemgs opened the pull request:
This commit is to support Tensorflow Lite 2.3.0 (TF2-Lite) additionally,
so that we support Tensorflow 2.x based network models (e.g., ASR) in
the Tizen 6.X platform.

**Changelog**

  * Version 3:
    * Removed unnecessary tflite_nnapi delegation (@jaeyun-jung)

  * Version 2:
    * Added a debian packaging for TF-2.3.0 in tensorflow2 (review.tizen.org)
      * https://review.tizen.org/gerrit/#/c/243189/ (Reviewed, Merged)
      * https://launchpad.net/~nnstreamer/+archive/ubuntu/ppa-build-test (Published)
  * Added unit-test with ssat

  * Version 1:
    * Added tflite2_support feature for meson build procedure
    * Added TF2-Lite statements in the .spec file for Tizen
    * Generate two libraries with "-DTFLITE_VERSION" flag from
      the tensor_filter_tensorflow_lite.cc file
    * Added nnstreamer-tensorflow2-lite package

Signed-off-by: Geunsik Lim <geunsik.lim@samsung.com>

leemgs commented:
@jaeyun-jung PTAL, Please review this PR.

jaeyun-jung commented on ext/nnstreamer/tensor_filter/meson.build:
```diff
@@ -134,6 +135,57 @@ if tflite_support_is_available
 endif
 
+if tflite2_support_is_available
+  filter_sub_tflite2_sources = ['tensor_filter_tensorflow_lite.cc']
```
later work: let's try to reduce duplicated script for tf-lite 1 & 2.

leemgs replied:
Yes. And, we need to talk about the maintenance policy to handle efficiently versions of TF in the near future as we have learned  at the Python package (e.g., Python 2.X , Python 3.X, and Python4.X)

helloahn reviewed (commented):
Is it available using TF 1.x and TF 2.x simultaneously? If it isn't, we have to make a new policy for `enable_tf_lite` and `enable_tf2_lite`, for instance, if 1.x is chosen, 2.x becomes disabled. IMO, because of the `protobuf`, `flatbuffers` we should consider the compatibility.

leemgs commented:
> Is it available using TF 1.x and TF 2.x simultaneously?

@helloahn , Good comment. Yes,  We are going to support both TF 1.x and 2.X  with a multiple library structure such as Python (V2, V3). 
* TF 1.X:   /usr/lib/tensorflow/lite
* TF 2.X:  /usr/lib/tensorflow2/lite

And, in the near future, we need to append how we do support a default framework (e.g., symbolic link with post-install) for compatibility of existing applications.

leemgs commented:
To reviewers, 

**Changelog**

  * Version 2:
    * Added a debian packaging for TF-2.3.0 in tensorflow2 (review.tizen.org)
      * https://review.tizen.org/gerrit/#/c/243189/ (Reviewed, Merged)
      * https://launchpad.net/~nnstreamer/+archive/ubuntu/ppa-build-test (Published)
    * Added unit-test with ssat

  * Version 1:
    * Added tflite2_support feature for meson build procedure
    * Added TF2-Lite statements in the .spec file for Tizen
    * Generate two libraries with "-DTFLITE_VERSION" flag from
      the tensor_filter_tensorflow_lite.cc file
    * Added nnstreamer-tensorflow2-lite package

myungjoo commented:
TF2-Lite support is not RELEASE CRITICAL.

Actually, this should NOT be included in 1.6.y releases. This should be included in 1.7.x (1.8.0 rc-x) releases.

jaeyun-jung commented on ext/nnstreamer/tensor_filter/meson.build:
```diff
@@ -134,6 +135,57 @@ if tflite_support_is_available
+
+  nnstreamer_filter_tflite2_deps = tflite_support_deps + [thread_dep, libdl_dep, glib_dep, gst_dep, nnstreamer_dep]
+
+  if tflite_nnapi_delegation_is_available
```
This is unnecessary (nnfw delegation in tf-lite subplugin)
I made a patch to remove this feature, plz delete this line.

leemgs replied:
Applied. Thanks.

leemgs commented:
UPDATED.

**Changelog**

  * Version 3:
    * Removed unnecessary tflite_nnapi delegation (@jaeyun-jung)

  * Version 2:
    * Added a debian packaging for TF-2.3.0 in tensorflow2 (review.tizen.org)
      * https://review.tizen.org/gerrit/#/c/243189/ (Reviewed, Merged)
      * https://launchpad.net/~nnstreamer/+archive/ubuntu/ppa-build-test (Published)
  * Added unit-test with ssat

  * Version 1:
    * Added tflite2_support feature for meson build procedure
    * Added TF2-Lite statements in the .spec file for Tizen
    * Generate two libraries with "-DTFLITE_VERSION" flag from
      the tensor_filter_tensorflow_lite.cc file
    * Added nnstreamer-tensorflow2-lite package

myungjoo commented on ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:
```diff
@@ -1101,7 +1103,11 @@ tflite_checkAvailability (accl_hw hw)
 static gchar filter_subplugin_tensorflow_lite[] = "tensorflow-lite";
+#else
+static gchar filter_subplugin_tensorflow_lite[] = "tensorflow2-lite";
+#endif
```
WHat if both tf2-lite and tf1-lite are installed simultaneously?

dongju-chae replied:
I think it's okay because even if they share the same source files, different .so files (and different names) will be created.

dongju-chae reviewed (approved):
Looks good!

myungjoo reviewed (approved):
LGTM.

Now, we need to update Single-API. You will need some careful design there.

wooksong commented on ext/nnstreamer/tensor_filter/meson.build:
```diff
@@ -134,6 +135,53 @@ if tflite_support_is_available
+    tflite2_compile_args += '-DTFLITE_FLOAT16=1'
+  endif
+  if cxx.has_type('kTfLiteComplex64', prefix : '#include <tensorflow2/contrib/lite/model.h>')
+    tflite2_compile_args += '-DTFLITE_COMPLEX64=1'
```
IMO, tflite2 supports all those data types. Therefore, in the case of tfilte2, we do not need theses cxx.has_type statements.
